### PR TITLE
make shell pop layout local

### PR DESCRIPTION
**Pull request by CeleritasCelery**
There are two commits here. The first adds new functionality to allow users to specify certain variables as "layout local" (meaning the the value will be bound to the current layout).

The second commit uses this new mechanism to make shell-pop's shells bound to the current layout. This behavior can be toggled with `layouts-enable-local-variables`.

### CODE REVIEW

File: layers/+spacemacs/spacemacs-layouts/funcs.el [hunk 1]
```diff
@@ -654,3 +654,32 @@ FRAME defaults to the current frame."
...
+
+

+;; layout local variables
+
+(defun spacemacs/make-variable-layout-local (&rest vars)
+  "Make variables become layout-local whenever they are set.
+Accepts a list of VARIABLE, DEFAULT-VALUE pairs.
+
+(spacemacs/make-variable-layout-local 'foo 1 'bar 2)"
```
Comment (nixmaniack) [inline layers/+spacemacs/spacemacs-layouts/funcs.el]: Looks like leftover function call?

**Comment by aaronjensen** [comment]
When using this, sometimes, after closing the pop shell, it changes my current buffer to that of another layout. 

I'm testing by opening two layouts using `SPC p l`, each with one window/buffer. 

I pop the shell in one, then switch to the other, then pop the shell in there. I think toggle back and forth and close/reopen the shells. Occasionally the buffer switches so that the buffer in Persp A is the one from Persp B

**Comment by CeleritasCelery** [comment]
@aaronjensen I am having a hard time reproducing your issue. Are you able to create dedicated reproduction steps? Also when you say "close/reopen" do you mean pop-in/pop-out or literally close exit the shells and restart new ones? 



> Occasionally the buffer switches so that the buffer in Persp A is the one from Persp B

I wonder if this is because we are not making `shell-pop-last-buffer` layout-local. would you try that and see if it remedies your issue?

**Comment by aaronjensen** [comment]
That helps. The repro steps are:

1. Open file in persp A
2. Pop shell
3. Ope

**Comment by aaronjensen** [comment]
That helps. The repro steps are:

1. Open file in persp A
1. Open file in persp B
1. Switch to A and pop shell
1. Switch to B and pop shell
1. Switch back to A and pop shell again (which should close the existing one)

After making `shell-pop-last-buffer` layout local, I don't get the wrong buffer, but I do get this:

```
Debugger entered--Lisp error: (wrong-type-argument window-live-p #<window 18>)
  #<subr select-window>(#<window 18> nil)
  ad-Advice-select-window(#<subr select-window> #<window 18>)
  apply(ad-Advice-select-window #<subr select-window> #<window 18>)
  #f(advice-wrapper :around #<subr select-window> ad-Advice-select-window)(#<window 18>)
  apply(#f(advice-wrapper :around #<subr select-window> ad-Advice-select-window) #<window 18>)
  select-window(#<window 18>)
  shell-pop-out()
  shell-pop(nil)
  (if (equal '(4) index) (ansi-term shell-pop-term-shell) (shell-pop--set-shell-type 'shell-pop-shell-type (list "ansi-term" (concat "*" (spacemacs//current-layout-name) "-" (if (file-remote-p default-directory) "remote-" "") "ansi-term" "*") (lambda nil (ansi-term shell-pop-term-shell)))) (shell-pop index))
  spacemacs/shell-pop-ansi-term(nil)
  funcall-interactively(spacemacs/shell-pop-ansi-term nil)
  call-interactively(spacemacs/shell-pop-ansi-term)
  (let ((shell (if (eq 'multi-term shell-default-shell) 'multiterm shell-default-shell))) (call-interactively (intern (format "spacemacs/shell-pop-%S" shell))))
  spacemacs/default-pop-shell()
  funcall-interactively(spacemacs/default-pop-shell)
  call-interactively(spacemacs/default-pop-shell nil nil)
  command-execute(spacemacs/default-pop-shell)
```

I get this even if I make `shell-pop-last-window` layout local...

**Comment by CeleritasCelery** [comment]
We get this error because when `persp-mode` switches to a new layout it creates new window objects in the same configuration as the old ones. The old window objects are not preserved so when we we try to select it we get that failure. I was able to fix this problem by updating `shell-pop-last-window` in `shell-pop-out-hook` using the code below
```lisp
(defun spacemacs//shell-pop-restore-window ()
  (unless (window-live-p shell-pop-last-window)
    (setq shell-pop-last-window (get-buffer-window shell-pop-last-buffer))))
(add-hook 'shell-pop-out-hook 'spacemacs//shell-pop-restore-window)
```
If this code fixes it for you as well I will push it to the PR branch.

**Comment by aaronjensen** [comment]
This seems to work great for me.

**Comment by CeleritasCelery** [comment]
@aaronjensen, can you reproduce this issue that I am seeing? 
1. have a perspective with two buffers in it (A and B)
2. pop in a shell from buffer A
3. go back to buffer A and delete it
4. return to the shell and pop out

rather then poping back to buffer B it tries to restore a blank version of buffer A. Are you seeing something similar?

**Comment by aaronjensen** [comment]
@CeleritasCelery Yes, I see the same.

**Comment by CeleritasCelery** [comment]
The delete buffer issue seems to be a `shell-pop` bug. But we might as well fix both of the issues at the same time.

**Comment by aaronjensen** [comment]
It all seems to work well for me.

**Comment by Compro-Prasad** [comment]
I think this seems ready. @syl20bnr

**Comment by d12frosted** [comment]
Hi everyone! I see some nice work here 😸 Gonna play around with your code.

@CeleritasCelery I know it's a little bit too much, but can you please do a rebase?

**Comment by CeleritasCelery** [comment]
@d12frosted 
rebased

**Comment by duianto** [comment]
Update request:
- Rebase and resolve the conflicts
- Update `changelog.develop`

Question:
Does this PR cover the behavior of the similar PR:
Add option to enable separate shells for each layout #9386

If yes, then we can be close that one.

**Comment by smile13241324** [comment]
@duianto this looks like it has been rebased and would be ready for review.
Could you have a look? I am still totally uneducated when it comes to layouts in spacemacs.

**Comment by duianto** [comment]
We'll close this PR, because the following PRs changes, have been applied to the develop branch:
Local shell pop #13556

**Pull request closed without merging**